### DISCUSSION TITLE
Timers editor in progress!

### DISCUSSION THREAD

**Discussion by dbuezas** [dbuezas/lgt8fx#164] [Show and tell]
Each time I need to do something non-trivial with timers, I have to spend hours with the datasheets. Hopefully there will be another way soon! :)

Visually it is extremely basic still. I may also visualise the behaviour of the outputs and the timer counter.

![image](https://user-images.githubusercontent.com/777196/119406200-bb8ad200-bce2-11eb-812b-ebbd5a80a5ff.png)

Internally, it computes the values of all registers, how they translate to behaviour and how they depend on each other. It works as a massive join of all register tables. As you select options, the incompatible ones go disabled and the ones that have no more alternatives get fixed.

If I can grasp them, I'll also add the lgt extra twists like dead time and shut down registers.

**Comment by dbuezas** [comment]
It's taking shape!
The only in-progress part now is the plot (I just put the screen shot of an experimental gsheet where I tested the timer data generating algorithm)

![Screen Shot 2021-05-30 at 15 06 15](https://user-images.githubusercontent.com/777196/120105296-a4bc0380-c158-11eb-95b8-8489cd634c59.png)

**Reply by dbuezas** [reply]
Oh, and it doesn't have dead time support yet.
There are already 16 internal tables per timer that are used by the constraint solver to find out `what can you combined with what`, `what everything means for each register`, and `how to plot it`

**Comment by LaZsolt** [comment]
I remember seeing something like that somewhere.
SMT32cube software can generate codes like this for STM32 microcontrollers.
But I found tools for AVR.
https://hackaday.com/2020/02/06/tool-writes-your-pwm-code-for-you/
or **CodeVisionAVR**

**Reply by dbuezas** [reply]
darn 😅... well, this is a webpage, and for lgt

**Comment by dbuezas** [comment]
The tool is finished and you can find it here by the way:

https://dbuezas.github.io/arduino-web-timers/
<img width="919" alt="image" src="https://user-images.githubusercontent.com/777196/213936398-a40ef532-bc4a-46a6-8f1a-a2e3e0a94ce0.png">
